File: web/src/lib/devicePair.ts
export const DEVICE_PAIR_URI_PREFIX = 'ultrasend://pair/';

export function devicePairUri(deviceId: string): string {
  return `${DEVICE_PAIR_URI_PREFIX}${deviceId}`;
}

export function parseDevicePairUri(text: string): string | null {
  const trimmed = text.trim();
  if (!trimmed.toLowerCase().startsWith(DEVICE_PAIR_URI_PREFIX)) return null;
  const rest = trimmed.slice(DEVICE_PAIR_URI_PREFIX.length).trim();
  const id = rest.split(/[?#]/)[0]?.trim() ?? '';
  return id.length > 0 ? id : null;
}

/** Accepts `ultrasend://pair/<id>` or a raw deviceId. */
export function normalizePeerDeviceId(raw: string): string | null {
  const trimmed = raw.trim();
  if (!trimmed) return null;
  return parseDevicePairUri(trimmed) ?? trimmed;
}

export function shortDeviceLabel(deviceId: string): string {
  if (deviceId.length <= 12) return deviceId;
  return `${deviceId.slice(0, 8)}…`;
}

File: app/lib/device_pair.dart
const kDevicePairUriPrefix = 'ultrasend://pair/';

String devicePairUri(String deviceId) => '$kDevicePairUriPrefix$deviceId';

String? parseDevicePairUri(String text) {
  final trimmed = text.trim();
  if (!trimmed.toLowerCase().startsWith(kDevicePairUriPrefix)) return null;
  final rest = trimmed.substring(kDevicePairUriPrefix.length).trim();
  final id = rest.split(RegExp(r'[?#]')).first.trim();
  return id.isEmpty ? null : id;
}

/// Accepts `ultrasend://pair/<id>` or a raw deviceId.
String? normalizePeerDeviceId(String raw) {
  final trimmed = raw.trim();
  if (trimmed.isEmpty) return null;
  return parseDevicePairUri(trimmed) ?? trimmed;
}

String shortDeviceLabel(String deviceId) {
  if (deviceId.length <= 12) return deviceId;
  return '${deviceId.substring(0, 8)}…';
}
